File: dicm2dckv/dicmstructdump/dckvapi.m
//
//  dckvapi.m
//  dicm2dckv
//
//  Created by jacquesfauquex on 2024-04-04.
//

/*potencialmente multivalue: AE AS AT CS DA DS DT FD FL IS LO OW PN SH SV TM UC UI UV*/
#include <Foundation/Foundation.h>

#include "dckvapi.h"

size_t dckvapi_fread(
                     void * __restrict __ptr,
                     size_t __size,
                     size_t __nitems,
                     FILE * __restrict __stream
                     )
{
   return fread(__ptr,__size,__nitems,__stream);
}

uint8 swapchar;


//returns true when 8 bytes were read
BOOL dckvapi_fread8(uint8_t *buffer, unsigned long *bytesReadRef)
{
   *bytesReadRef=fread(buffer, 1, 8, stdin);
   if (ferror(stdin)){
      E("%s","stdin error");
      return false;
   }
   
   if (*bytesReadRef==8){
      swapchar=*buffer;
      *buffer=*(buffer+1);
      *(buffer+1)=swapchar;
      swapchar=*(buffer+2);
      *(buffer+2)=*(buffer+3);
      *(buffer+3)=swapchar;
   }
   else
   {
      *buffer=0xFF;
      *(buffer+1)=0xFF;
      *(buffer+2)=0xFF;
      *(buffer+3)=0xFF;
   }
   return true;
}

const char *space=" ";
const char *backslash = "\\";

#pragma mark obligatorios formales
bool createtx(
   const char * dstdir,
   uint8_t    * vbuf,
   uint64 *soloc,         // offset in valbyes for sop class
   uint16 *solen,         // length in valbyes for sop class
   uint16 *soidx,         // index in const char *scstr[]
   uint64 *siloc,         // offset in valbyes for sop instance uid
   uint16 *silen,         // length in valbyes for sop instance uid
   uint64 *stloc,         // offset in valbyes for transfer syntax
   uint16 *stlen,         // length in valbyes for transfer syntax
   uint16 *stidx,         // index in const char *csstr[]
   sint16 *siidx          // SOPinstance index
){
   I("#%d",*siidx);
   printf("     144 %s\n","00020001 OB 0000 {156,2}");
   return true;
}
bool committx(sint16 *siidx){
   return closetx(siidx);
}
bool closetx(sint16 *siidx){
   I("!#%d",*siidx);
   (*siidx)++;
   return true;
}


#pragma mark - parseo y agregado

bool appendkv(
              uint8_t            *kbuf,
              unsigned long      kloc,
              BOOL               vlenisl,
              enum kvVRcategory  vrcat,
              unsigned long long vloc,
              unsigned long      vlen,
              BOOL               fromStdin,
              uint8_t            *vbuf
              )
{
   //loc+datasetspace+itemspace+item
   switch (kloc) {
      case 0: printf("%8lld ",vloc);
         break;
      case 8:
      {
         printf("%8lld         %02X%02X%02X%02X ",vloc, kbuf[kloc-4], kbuf[kloc-3], kbuf[kloc-2], kbuf[kloc-1]);
      }
         break;
      default:
      {
         printf("%8lld%*s%02X%02X%02X%02X ",vloc,(int)(((kloc/8)*9)+8),space, kbuf[kloc-4], kbuf[kloc-3], kbuf[kloc-2], kbuf[kloc-1]);
      }
   }
   
   
   
   //uint64* attruint64=(uint64*) kbuf+kloc;
   //printf("%*s%016llX\n",kloc+kloc+1,space, CFSwapInt64(*attruint64));
   uint32* t=(uint32*)kbuf+(kloc/4);
   uint32* i=0;
   if (kloc > 0) i=(uint32*) kbuf+(kloc/4)-4;
   uint16* l=(uint16*) kbuf+((kloc/2)+3);
   uint8 v=*(kbuf+(kloc+4));
   uint8 r=*(kbuf+(kloc+5));
      
   if (vlenisl)
   {
#pragma mark long length
      if (vlen==0xFFFFFFFF)
      {
         unsigned long long ulllen=0;
         if (fromStdin)
         {
            size_t bytesread=0xFFFE;
            while (bytesread==0xFFFE)
            {
               bytesread=fread(vbuf,1,0xFFFE,stdin);
               ulllen+=bytesread;
            }
         }
         else if (!fromStdin)
         {
            //contents already in vbuf
         }
         switch (vrcat) {
             
            case kvAl://AccessionNumberIssuer local 00080051.00400031
            case kvAu://AccessionNumberIssuer universal 00080051.00400032
            case kvTL://UC

            case kvTU://UR
            
            case kved://OB Encapsulated​Document 00420011
            case kvfo://OV Extended​Offset​Table fragments offset 7FE00001
            case kvfl://OV Extended​Offset​TableLengths fragments offset 7FE00002
            case kvft://UV Encapsulated​Pixel​Data​Value​Total​Length 7FE00003
            case kv01://OB OD OF OL OV OW SV UV
               
            case kvUN:
               printf("%08X %c%c %04X {%llu,%llu}\n",CFSwapInt32(*t),v,r,*l,vloc+12,ulllen);
               break;
               
            case kvSA://SQ head
            {
               printf("%08X00000000\n", CFSwapInt32(*t));
            }break;

            default: return false;
         }

      }
      else
      {
#pragma mark short length
         
         if (fromStdin && vlen)
         {
            unsigned long ll=vlen;
            while (ll>0xFFFE)
            {
               if (fread(vbuf,1,0xFFFE,stdin)!=0xFFFE) return false;
               ll-=0xFFFE;
            }
            if (ll && (!feof(stdin)) && (fread(vbuf,1,ll,stdin)!=ll)) return false;
         }
         else if (!fromStdin)
         {
            //contents already in vbuf
         }
         
         
         switch (vrcat) {
             
            case kvAl://AccessionNumberIssuer local 00080051.00400031
            case kvAu://AccessionNumberIssuer universal 00080051.00400032
            case kvTL://UC

            case kvTU://UR
               
            case kv01://OB OD OF OL OV OW SV UV
               
            case kvUN:
               printf("%08X %c%c %04X {%llu,%lu}\n",CFSwapInt32(*t),v,r,*l,vloc+12,vlen);
               break;
               
            case kvSA://SQ head
            {
               printf("%08X00000000\n", CFSwapInt32(*t));
            }break;

            default: return false;
         }
      }
   }
   else //vl
   {
      if (fromStdin && vlen && (fread(vbuf,1,vlen,stdin)!=vlen)) return false;
      
      switch (vrcat) {
         case kvFD://floating point double
         {
            printf("%08X %c%c %04X",CFSwapInt32(*t),v,r,*l);
            if (vlen > 0)
            {
               double *v=(double*)vbuf;
               printf(" (");
               for (uint16 idx=0; idx<(vlen>>3); idx++)
               {
                  printf(" %f",v[idx]);
               }
               printf(" )");
            }
            printf("\n");
         }break;
            
         case kvFL://floating point single
         {
            printf("%08X %c%c %04X",CFSwapInt32(*t),v,r,*l);
            if (vlen > 0)
            {
               float *v=(float*)vbuf;
               printf(" (");
               for (uint16 idx=0; idx<(vlen>>2); idx++)
               {
                  printf(" %f",v[idx]);
               }
               printf(" )");
            }
            printf("\n");
         }break;
            
         case kvSL://signed long
         {
            printf("%08X %c%c %04X",CFSwapInt32(*t),v,r,*l);
            if (vlen > 0)
            {
               sint32 *v=(sint32*)vbuf;
               printf(" (");
               for (uint16 idx=0; idx<(vlen>>2); idx++)
               {
                  printf(" %d",v[idx]);
               }
               printf(" )");
            }
            printf("\n");
         }break;
            
         case kvSS://signed short
         {
            printf("%08X %c%c %04X",CFSwapInt32(*t),v,r,*l);
            if (vlen > 0)
            {
               sint16 *v=(sint16*)vbuf;
               printf(" (");
               for (uint16 idx=0; idx<(vlen>>1); idx++)
               {
                  printf(" %hd",v[idx]);
               }
               printf(" )");
            }
            printf("\n");
         }break;
            
         case kvUL://unsigned long
         {
            printf("%08X %c%c %04X",CFSwapInt32(*t),v,r,*l);
            if (vlen > 0)
            {
               uint32 *v=(uint32*)vbuf;
               printf(" (");
               for (uint16 idx=0; idx<(vlen>>2); idx++)
               {
                  printf(" %u",v[idx]);
               }
               printf(" )");
            }
            printf("\n");
         }break;
            
         case kvUS://unsigned short
         {
            printf("%08X %c%c %04X",CFSwapInt32(*t),v,r,*l);
            if (vlen > 0)
            {
               uint16 *v=(uint16*)vbuf;
               printf(" (");
               for (uint16 idx=0; idx<(vlen>>1); idx++)
               {
                  printf(" %hu",v[idx]);
               }
               printf(" )");
            }
            printf("\n");
         }break;
            
         case kvAT://attribute tag
         {
            printf("%08X %c%c %04X",CFSwapInt32(*t),v,r,*l);
            if (vlen > 0)
            {
               uint16 *v=(uint16*)vbuf;
               printf(" (");
               for (uint16 idx=0; idx<(vlen>>2); idx+=2)
               {
                  printf(" %04x%04x",v[idx],v[idx+1]);
               }
               printf(" )");
            }
            printf("\n");
         }break;
            
         case kvII://SOPInstanceUID
         case kvIE://StudyInstanceUID
         case kvIS://SeriesInstanceUID
         case kvUI://unique ID
         {
            printf("%08X %c%c %04X",CFSwapInt32(*t),v,r,*l);
            if (vlen > 0)
            {
               printf(" (");
               vbuf[vlen]=0x0;//terminate even uid lists
               char *token = strtok((char *)vbuf, backslash);
               while( token != NULL ) {
                     printf( " %s", token );
                     token = strtok(NULL, backslash);
               }
               printf(" )");
            }
            printf("\n");
         }break;
            
         case kvEd://StudyDate
         case kvTP:
            
         case kvSm://Modality
         case kvAt://AccessionNumberType
         case kvIs://SeriesNumber
         case kvIi://InstanceNumber
         case kvIa://AcquisitionNumber
         case kvTA://AE DS IS CS
            
         case kvdn://ST  DocumentTitle 00420010
         case kvHC://HL7InstanceIdentifier
         case kvEi://StudyID
         case kvAn://AccessionNumber
         case kvIN://InstitutionName
         case kvTS://LO LT SH ST
            
         case kvPN:
         {
            printf("%08X %c%c %04X",CFSwapInt32(*t),v,r,*l);
            if (vlen > 0)
            {
               printf(" (");
               vbuf[vlen]=0x0;//terminate even uid lists
               char *token = strtok((char *)vbuf, backslash);
               while( token != NULL ) {
                     printf( " «%s»", token );
                     token = strtok(NULL, backslash);
               }
               printf(" )");
            }
            printf("\n");
         }break;
            
         case kvIA://item head
         {
            printf("%08X %c%c %04X\n",CFSwapInt32(*t),v,r,*l);
         }break;
            
         case kvIZ://item tail
         {
            printf("%08X %c%c %04X\n",CFSwapInt32(*t),v,r,*l);
         }break;
            
         case kvSZ://SQ tail
         {
            printf("%08XFFFFFFFF\n", CFSwapInt32(*t));
         }break;

         default: return false;
      }
   }

   return true;
}

File: dicm2dckv/dicmstructdump/dckvapi.c
// project: dicm2dckv
// target: dicmstructdump
// file: dckvapi.c
// created by jacquesfauquex on 2024-04-04.


#include "dckvapi.h"

extern char *DICMbuf;
extern u64 DICMidx;
extern s16 siidx;
extern uint8_t *kbuf;

#pragma mark - read

static u64 bytesreceived;
bool _DKVfread(u32 bytesaskedfor)
{
   bytesreceived=fread(DICMbuf+DICMidx,1,bytesaskedfor,stdin);
   if (bytesreceived>0xFFFFFFFF)return 0;
   DICMidx+=bytesreceived;
   return (bytesaskedfor==bytesreceived);
}


//returns true when 8(+4) bytes were read
bool _DKVfreadattr(u8 kloc)
{
   if (fread(DICMbuf+DICMidx,1,8,stdin)!=8)
   {
      if (ferror(stdin)) E("%s","stdin error");
      return false;
   }
   
   //group LE>BE
   kbuf[kloc]=DICMbuf[DICMidx+1];
   kbuf[kloc+1]=DICMbuf[DICMidx];
   //element LE>BE
   kbuf[kloc+2]=DICMbuf[DICMidx+3];
   kbuf[kloc+3]=DICMbuf[DICMidx+2];
   //vr vl copied (LE)
   kbuf[kloc+4]=DICMbuf[DICMidx+4];
   kbuf[kloc+5]=DICMbuf[DICMidx+5];
   kbuf[kloc+6]=DICMbuf[DICMidx+6];
   kbuf[kloc+7]=DICMbuf[DICMidx+7];

   switch ((DICMbuf[DICMidx+5]<<8)|(DICMbuf[DICMidx+4])) {
      case OB://other byte
      case OW://other word
      case OD://other double
      case OF://other float
      case OL://other long
      case SV://signed 64-bit very long
      case OV://other 64-bit very long
      case UV://unsigned 64-bit very long
      case UC://unlimited characters
      case UT://unlimited text
      case UR://universal resrcurl identifier/locator
      case SQ://sequence
      {
         DICMidx+=8;
         if (fread(DICMbuf+DICMidx,1,4,stdin)!=4)
         {
            if (ferror(stdin)) E("%s","stdin error");
            return false;
         }
         memcpy(kbuf+kloc+8, DICMbuf+DICMidx, 4);
         DICMidx+=4;
      }break;
      default:
      {
         //IA,IZ,SZ require postprocessing in dicm2dckv
         DICMidx+=8;
         memcpy(kbuf+kloc+8, DICMbuf+DICMidx-2, 2);
         kbuf[kloc+10]=0;
         kbuf[kloc+11]=0;
      }break;
   }
   
   return true;
}

const char *space=" ";

#pragma mark - instance transactions

bool _DKVcreate(
   u64 soloc,         // offset in valbyes for sop class
   u16 solen,         // length in valbyes for sop class
   u16 soidx,         // index in const char *scstr[]
   u64 siloc,         // offset in valbyes for sop instance uid
   u16 silen,         // length in valbyes for sop instance uid
   u64 stloc,         // offset in valbyes for transfer syntax
   u16 stlen,         // length in valbyes for transfer syntax
   u16 stidx          // index in const char *csstr[]
)
{
   D("dump create #%d",siidx);
   printf("     144 %s\n","00020001 OB 0000 {156,2}");
   printf("%8llu 00020002 UI 0000 \"%s\" [%hu]\n",soloc,DICMbuf+soloc,soidx);
   printf("%8llu 00020003 UI 0000 \"%s\"\n",siloc,DICMbuf+siloc);
   printf("%8llu 00020010 UI 0000 \"%s\" [%hu]\n",stloc,DICMbuf+stloc,stidx);
   return true;
}
bool _DKVcommit(void)
{
   FILE *fileptr=fopen("dicmstructdump.dcm", "w");
   if (fileptr == NULL) return false;
   if (fwrite(DICMbuf ,1, DICMidx , fileptr)!=DICMidx) return false;
   fclose(fileptr);
   return _DKVclose();
}
bool _DKVclose(void)
{
   I("!#%d", siidx);
   return true;
}

#pragma mark - write

bool _DKVappend(int kloc,enum kvVRcategory vrcat,u32 vlen)
{
   switch (vrcat) {
      case kvSA: printf("%8llu%*s%02X%02X%02X%02X+\n",DICMidx-12,kloc+kloc+(kloc!=0),space, kbuf[kloc],kbuf[kloc+1],kbuf[kloc+2],kbuf[kloc+3]);break;
      case kvSZ: printf("%8llu%*s%02X%02X%02X%02X~\n",DICMidx-8,kloc+kloc+(kloc!=0),space, kbuf[kloc],kbuf[kloc+1],kbuf[kloc+2],kbuf[kloc+3]);break;
      case kvIA: printf("%8llu %*s%02X%02X%02X%02X+\n",DICMidx-8,kloc+kloc-8,space,kbuf[kloc-4],kbuf[kloc-3],kbuf[kloc-2],kbuf[kloc-1]);break;
      case kvIZ: printf("%8llu %*s%02X%02X%02X%02X~\n",DICMidx-8,kloc+kloc-8,space,kbuf[kloc-4],kbuf[kloc-3],kbuf[kloc-2],kbuf[kloc-1]);break;
      case kv01://OB OD OF OL OV OW SV UV
      case kvsdocument://OB Encapsulated​Document 00420011 xml cda o pdf
      case kvnativeOB: //OB 0x7FE00010
      case kvnativeOW: //OW 0x7FE00010
      case kvnativeOD: //OD 0x7FE00009
      case kvnativeOF: //OF 0x7FE00008
      case kvnativeOC: //OB 0x7E000010
      case kvframesOB: //OB 0x7E000010
      case kvframesOC: //OB 0x7E000010
      case kvfo://OV Extended​Offset​Table fragments offset 7FE00001
      case kvfl://OV Extended​Offset​TableLengths fragments offset 7FE00002
      case kvft://UV Encapsulated​Pixel​Data​Value​Total​Length 7FE00003
      case kvUN: {
         printf("%8llu%*s%02X%02X%02X%02X %c%c %04X ",DICMidx-12,kloc+kloc+(kloc!=0),space, kbuf[kloc],kbuf[kloc+1],kbuf[kloc+2],kbuf[kloc+3],kbuf[kloc+4],kbuf[kloc+5],kbuf[kloc+6] + (kbuf[kloc+7] << 8));
         if (!_DKVfread(vlen)) return false;
         printf("{%llu,%u}\n",DICMidx-vlen,vlen);
      }break;
      case kvTL://UC
      case kveal://UT AccessionNumberIssuer local 00080051.00400031
      case kveau://UT AccessionNumberIssuer universal 00080051.00400032
      case kvTU: { //UR
         printf("%8llu%*s%02X%02X%02X%02X %c%c %04X ",DICMidx-12,kloc+kloc+(kloc!=0),space, kbuf[kloc],kbuf[kloc+1],kbuf[kloc+2],kbuf[kloc+3],kbuf[kloc+4],kbuf[kloc+5],kbuf[kloc+6] + (kbuf[kloc+7] << 8));
         if (vlen>0)
         {
            //charset -> utf-8
            u32 repidx=kbuf[kloc+6] + (kbuf[kloc+7] << 8);
            u32 charstart=(u32)DICMidx;
            u32 utf8length=0;
            if (!_DKVfread(vlen)) return false;
            utf8(repidx,DICMbuf,charstart,vlen,DICMbuf,(u32)DICMidx,&utf8length);
            printf( "\"%.*s\"\n", utf8length,DICMbuf+DICMidx );
         }
      } break;
      case kvFD: { //floating point double
         printf("%8llu%*s%02X%02X%02X%02X %c%c %04X ",DICMidx-8,kloc+kloc+(kloc!=0),space, kbuf[kloc],kbuf[kloc+1],kbuf[kloc+2],kbuf[kloc+3],kbuf[kloc+4],kbuf[kloc+5],kbuf[kloc+6] + (kbuf[kloc+7] << 8));
         if (vlen > 0)
         {
            printf("(");
            double d;
            if (!_DKVfread(vlen)) return false;
            for (u16 idx=DICMidx-vlen; idx<DICMidx; idx+=8)
            {
               memcpy(&d, DICMbuf+idx, 8);
               printf(" %f",d);
            }
            printf(" )");
         }
         printf("\n");
      }break;
      case kvFL: { //floating point single
         printf("%8llu%*s%02X%02X%02X%02X %c%c %04X ",DICMidx-8,kloc+kloc+(kloc!=0),space, kbuf[kloc],kbuf[kloc+1],kbuf[kloc+2],kbuf[kloc+3],kbuf[kloc+4],kbuf[kloc+5],kbuf[kloc+6] + (kbuf[kloc+7] << 8));
         if (vlen > 0)
         {
            printf("(");
            float f;
            if (!_DKVfread(vlen)) return false;
            for (u16 idx=DICMidx-vlen; idx<DICMidx; idx+=4)
            {
               memcpy(&f, DICMbuf+idx, 4);
               printf(" %f",f);
            }
            printf(" )");
         }
         printf("\n");
      }break;
      case kvSL: { //signed long
         printf("%8llu%*s%02X%02X%02X%02X %c%c %04X ",DICMidx-8,kloc+kloc+(kloc!=0),space, kbuf[kloc],kbuf[kloc+1],kbuf[kloc+2],kbuf[kloc+3],kbuf[kloc+4],kbuf[kloc+5],kbuf[kloc+6] + (kbuf[kloc+7] << 8));
         if (vlen > 0)
         {
            printf("(");
            s32 s4B;
            if (!_DKVfread(vlen)) return false;
            for (u16 idx=DICMidx-vlen; idx<DICMidx; idx+=4)
            {
               memcpy(&s4B, DICMbuf+idx, 4);
               printf(" %d",s4B);
            }
            printf(" )");
         }
         printf("\n");
      }break;
      case kvSS: { //signed short
         printf("%8llu%*s%02X%02X%02X%02X %c%c %04X ",DICMidx-8,kloc+kloc+(kloc!=0),space, kbuf[kloc],kbuf[kloc+1],kbuf[kloc+2],kbuf[kloc+3],kbuf[kloc+4],kbuf[kloc+5],kbuf[kloc+6] + (kbuf[kloc+7] << 8));
         if (vlen > 0)
         {
            printf("(");
            s16 s2B;
            if (!_DKVfread(vlen)) return false;
            for (u16 idx=DICMidx-vlen; idx<DICMidx; idx+=2)
            {
               memcpy(&s2B, DICMbuf+idx, 2);
               printf(" %hd",s2B);
            }
            printf(" )");
         }
         printf("\n");
      }break;
      case kvUL: { //unsigned long
         printf("%8llu%*s%02X%02X%02X%02X %c%c %04X ",DICMidx-8,kloc+kloc+(kloc!=0),space, kbuf[kloc],kbuf[kloc+1],kbuf[kloc+2],kbuf[kloc+3],kbuf[kloc+4],kbuf[kloc+5],kbuf[kloc+6] + (kbuf[kloc+7] << 8));
         if (vlen > 0)
         {
            printf("(");
            u32 u4B;
            if (!_DKVfread(vlen)) return false;
            for (u16 idx=DICMidx-vlen; idx<DICMidx; idx+=4)
            {
               memcpy(&u4B, DICMbuf+idx, 4);
               printf(" %u",u4B);
            }
            printf(" )");
         }
         printf("\n");
      }break;
      case kvUS://unsigned short
      case kvspp://7 US
      case kvrows://8 US
      case kvcols://9 US
      case kvalloc://10 US
      case kvstored://11 US
      case kvhigh://12 US
      case kvpixrep://13 US
      case kvplanar: { //14 US
         printf("%8llu%*s%02X%02X%02X%02X %c%c %04X ",DICMidx-8,kloc+kloc+(kloc!=0),space, kbuf[kloc],kbuf[kloc+1],kbuf[kloc+2],kbuf[kloc+3],kbuf[kloc+4],kbuf[kloc+5],kbuf[kloc+6] + (kbuf[kloc+7] << 8));
         if (vlen > 0)
         {
            printf("(");
            u16 u2B;
            if (!_DKVfread(vlen)) return false;
            for (u16 idx=DICMidx-vlen; idx<DICMidx; idx+=2)
            {
               memcpy(&u2B, DICMbuf+idx, 2);
               printf(" %hu",u2B);
            }
            printf(" )");
         }
         printf("\n");
      }break;
      case kvAT: { //attribute tag
         printf("%8llu%*s%02X%02X%02X%02X %c%c %04X ",DICMidx-8,kloc+kloc+(kloc!=0),space, kbuf[kloc],kbuf[kloc+1],kbuf[kloc+2],kbuf[kloc+3],kbuf[kloc+4],kbuf[kloc+5],kbuf[kloc+6] + (kbuf[kloc+7] << 8));
         if (vlen > 0)
         {
            printf("(");
            if (!_DKVfread(vlen)) return false;
            for (u16 idx=DICMidx-vlen; idx<DICMidx; idx+=2)
            {
               printf(" %04x%04x",*DICMbuf+idx,*DICMbuf+idx+1);
            }
            printf(" )");
         }
         printf("\n");
      }break;
      case kvUI://unique ID
      case kveuid://StudyInstanceUID
      case kvsuid://SeriesInstanceUID
      case kviuid://SOPInstanceUID
      case kvpuid://00080019 PyramidUID
      case kvcuid://SOP​Instance​UID​Of​Concatenation​Source
      case kvTP:
      case kvpbirth://Patient birthdate
      case kvedate://StudyDate
      case kvsdate://SeriesDate
      case kvstime://SeriesTime
      case kvTA://AE DS IS CS
      case kvpsex://CS patient sex
      case kveat://AccessionNumberType
      case kvsmod://Modality
      case kvitype://CS 00080008 ImageType
      case kvphotocode://CS
      case kvsnumber://SeriesNumber
      case kvianumber://AcquisitionNumber
      case kvinumber://InstanceNumber
      case kvscdaid:{ //ST HL7InstanceIdentifier 0040E001  root^extension
        printf("%8llu%*s%02X%02X%02X%02X %c%c %04X ",DICMidx-8,kloc+kloc+(kloc!=0),space, kbuf[kloc],kbuf[kloc+1],kbuf[kloc+2],kbuf[kloc+3],kbuf[kloc+4],kbuf[kloc+5],kbuf[kloc+6] + (kbuf[kloc+7] << 8));
        if (vlen > 0)
        {
           if (!_DKVfread(vlen)) return false;
           printf( "\"%.*s\"\n", vlen,DICMbuf+DICMidx-vlen );
        }
        else printf("\"\"\n");
     } break;
      //charset
      case kvTS://LO LT SH ST
      case kvpay://LO insurance
      case kvpide://SH patient id extension
      case kvpidr://LO patient id root issuer
      case kvimg://InstitutionName
      case kvedesc://LO Study name
      case kveid://StudyID
      case kvean://AccessionNumber
      case kvecode://SQ/SH Study code 00080100,00080102
      case kvsdesc://LO Series name
      case kvsdoctitle://ST  DocumentTitle 00420010
      case kvicomment://LO
      case kvPN:
      case kvpname://PN patient name
      case kvcda://PN CDA
      case kvref://PN referring
      case kvreq: { //PN requesting
         printf("%8llu%*s%02X%02X%02X%02X %c%c %04X ",DICMidx-8,kloc+kloc+(kloc!=0),space, kbuf[kloc],kbuf[kloc+1],kbuf[kloc+2],kbuf[kloc+3],kbuf[kloc+4],kbuf[kloc+5],kbuf[kloc+6] + (kbuf[kloc+7] << 8));
         if (vlen > 0)
         {
            //charset -> utf-8
            u32 repidx=kbuf[kloc+6] + (kbuf[kloc+7] << 8);
            u32 charstart=(u32)DICMidx;
            u32 utf8length=0;
            if (!_DKVfread(vlen)) return false;
            utf8(repidx,DICMbuf,charstart,vlen,DICMbuf,(u32)DICMidx,&utf8length);
            printf( "\"%.*s\"\n", utf8length,DICMbuf+DICMidx );
         }
         else printf("\"\"\n");
      } break;
         
      default: return false;
   
   }
   return true;
}
